pages/index.js

```diff
@@ -1,107 +1,233 @@
 import React from 'react'
 import Layout from '../components/Layout'
-import NodeView from '../components/NodeView'
 import fetch from 'isomorphic-unfetch'
 import input from '../input'
 
+import dynamic from 'next/dynamic'
+const JSONView = dynamic(
+  import('../components/JSONView'),
+  { ssr: false }
+)
+const TreeBeard = dynamic(
+  import('../components/TreeBeard'),
+  { ssr: false }
+)
+
 class Index extends React.Component {
 
   constructor(props) {
     super(props)
 
+    this.state = {
+      loading: false,
+      host: 'https://openfisca.mes-aides.gouv.fr',
+      source: '',
+      result: {},
+      root: '',
+      treeData: [],
+      situation: {},
+    }
+
+    this.fetchSource = this.fetchSource.bind(this);
+    this.fetchTrace = this.fetchTrace.bind(this);
     this.handleHostChange = this.handleHostChange.bind(this);
     this.handleRootCalculationChange = this.handleRootCalculationChange.bind(this);
-    this.handleTextAreaChange = this.handleTextAreaChange.bind(this);
+    this.handleSourceChange = this.handleSourceChange.bind(this);
     this.hasTrace = this.hasTrace.bind(this);
+  }
 
-    this.state = {
-      host: 'https://fr.openfisca.org/api/v21',//'http://127.0.0.1:2000',
-      situation: JSON.stringify(input, null, 2),
-      result: '',
-      root: ''
+  componentDidMount() {
+
+    const queryString = window.document.location.search.replace('?', '') || ''
+
+    const params = queryString.split('&').reduce((params, tuple) => {
+      const pair = tuple.split('=')
+
+      params[pair[0]] = decodeURI(pair[1])
+      return params
+    }, {})
+
+    let newState = {}
+    if (params.host) {
+      newState = Object.assign(newState, { host: params.host })
+    }
+    if (params.source) {
+      newState = Object.assign(newState, { source: params.source })
     }
-  }
 
-  hasTrace() {
-    return this.state.result && this.state.result.requestedCalculations
+    if (newState.hasOwnProperty('host') || newState.hasOwnProperty('source')) {
+      this.setState(newState)
+    }
   }
 
-  handleHostChange(event) {
-    this.setState({ host: event.target.value })
+  componentDidUpdate(prevProps, prevState) {
+    if (this.state.host !== prevState.host || this.state.source !== prevState.source) {
+      this.fetchSource()
```

Comment:
Shouldn't this line `return` to avoid fetching the trace to the incorrect source (as fetch source is async).